### ISSUE TITLE
Should macros import source implementations of their own package?

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/language#3474]
This is probably more a discussion regarding lints around macros.

In particular I was wondering:
Since macros will now enable generated code to add imports, would it be considered a bad idea if a macro imported its own source files?

For exemple a macro in a package named "A" could generate code which imports `import "package:a/src/something.dart"`

The reasoning is, I find it common that code generators want to use implementation details that should not be part of the public API of a package.

At the moment, the way I end up doing it is exporting those "private" APIs, but marking them as `@internal`, and ignoring that lint in generated code. But I was wondering if macros could enable solving that more officially

**Comment by lrhn** [comment]
Personal opinion: that would be reasonable.

The code that depends on the `lib/src/` located file is "from" the same package, by being generated by that package. It's reasonable to have that code depend on implementation details of other code on the same package.

Adding such an import is not allowing other packages to directly depend on those implementation details. Probably make sure to import it with a random prefix, if the macro system doesn't do that automatically, just to be absolutely sure someone doesn't add an accidental dependency on an imported name.

**Comment by rrousselGit** [comment]
I was thinking among those lines too.

I'm not necessarily fond of that /gendep, but I'd assume we'd need some sort of convention to false positives with the linter

**Comment by jakemac53** [comment]
Yes I agree that this is perfectly acceptable to do 👍. You will want to be careful to not expose any types in the generated API which are not public though, as that could end up forcing users to import those private libraries to use those types.

**Comment by jakemac53** [comment]
And also yes, this lint will probably always want to ignore macro augmentation files. Other lints may also want to do that, it might be worth auditing them at some point to see which should apply and which should not?

I don't think we want to globally disable them all because it would allow bad patterns to proliferate in generated code. At least the macro author should want to see any lints.

**Comment by rrousselGit** [comment]
In this scenario, I was thinking that all macros would start their generated file with `// generated by macro @macro from `package:foo`

Then the linter could silence `@internal` misuses specifically for `package:foo`,  but still warn about any other warning from different files

**Comment by jakemac53** [comment]
The analyzer might be able to just directly expose information about what macro generated a given declaration?

You don't have direct access to the file (there are many macros contributing to it), so you couldn't for instance add  your own comment to the top of the file easily. You can output comments above certain declarations, but I think it probably makes the most sense for the analyzer to just expose this info. Internally it probably has it already.

cc @scheglov

**Comment by scheglov** [comment]
Yes, we can do this.
I experimented with implementing it a bit.
For now just the reference to the macro class as `ClassElement`.
Theoretically we know more - which macro application annotation produced every macro generated element.
Don't know if such level of details would be useful.
```
library
  imports
    package:test/a.dart
  definingUnit
    classes
      class X @37
  augmentationImports
    package:test/test.macro.dart
      macroGeneratedCode
---
library augment 'test.dart';

import 'package:test/a.dart' as prefix0;

@prefix0.AddClassB()
class A {}

class B {}
---
      imports
        package:test/a.dart as prefix0 @62
      definingUnit
        classes
          class A @99
            generatedByMacro: package:test/a.dart::@class::AddClassA
          class B @111
            generatedByMacro: package:test/a.dart::@class::AddClassB
```

**Comment by srawlins** [comment]
I think this is dead.
